File: exercises/subsets_II_90.java
import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;
class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        List<List<Integer>> subsets = new ArrayList<>();
        List<Integer> subset = new ArrayList<>();
        Arrays.sort(nums);
        f(0, nums, subset, subsets);
        return subsets;
    }
    private void f(int index, int[] nums, List<Integer> subset, List<List<Integer>> subsets) {
        subsets.add(new ArrayList<>(subset));

        for (int i = index; i < nums.length; i++) {
            if (i != index && nums[i] == nums[i - 1]) continue;
            subset.add(nums[i]);
            f(i + 1, nums, subset, subsets);
            subset.remove(subset.size() - 1);
        }
    }
}

File: exercises/subsets_II_90.ts
function subsetsWithDup(nums: number[]): number[][] {
	const subsets: number[][] = []
	nums.sort((a,b) => a - b)

	function recurse(subset: number[], start: number) {
		subsets.push([...subset])
		for (let i = start; i < nums.length; i++) {
			if (i != start && nums[i] == nums[i-1]) continue
			subset.push(nums[i])
			recurse(subset, i + 1)
			subset.pop()
		}
	}

	recurse([], 0)
	return subsets
};

